hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/TokenUpdateLogic.java

```diff
@@ -0,0 +1,239 @@
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.CURRENT_TREASURY_STILL_OWNS_NFTS;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_EXPIRATION_TIME;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_ID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_IS_IMMUTABLE;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_IS_PAUSED;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_WAS_DELETED;
+
+import com.hedera.services.context.SideEffectsTracker;
+import com.hedera.services.exceptions.InvalidTransactionException;
+import com.hedera.services.ledger.SigImpactHistorian;
+import com.hedera.services.state.enums.TokenType;
+import com.hedera.services.state.merkle.MerkleToken;
+import com.hedera.services.store.contracts.WorldLedgers;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.NftId;
+import com.hedera.services.store.tokens.HederaTokenStore;
+import com.hedera.services.store.tokens.annotations.AreTreasuryWildcardsEnabled;
+import com.hedera.services.txns.util.TokenUpdateValidator;
+import com.hedera.services.txns.validation.OptionValidator;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TokenID;
+import com.hederahashgraph.api.proto.java.TokenUpdateTransactionBody;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+import java.util.Optional;
+import javax.inject.Inject;
+
+public class TokenUpdateLogic {
+    private final OptionValidator validator;
+    private final HederaTokenStore store;
+    private final WorldLedgers ledger;
+    private final SideEffectsTracker sideEffectsTracker;
+    private final SigImpactHistorian sigImpactHistorian;
+    boolean allowChangedTreasuryToOwnNfts;
+
+    @Inject
+    public TokenUpdateLogic(
+            final @AreTreasuryWildcardsEnabled boolean allowChangedTreasuryToOwnNfts,
+            OptionValidator validator,
+            HederaTokenStore tokenStore,
+            WorldLedgers ledger,
+            SideEffectsTracker sideEffectsTracker,
+            SigImpactHistorian sigImpactHistorian) {
+        this.validator = validator;
+        this.store = tokenStore;
+        this.ledger = ledger;
+        this.sideEffectsTracker = sideEffectsTracker;
+        this.sigImpactHistorian = sigImpactHistorian;
+        this.allowChangedTreasuryToOwnNfts = allowChangedTreasuryToOwnNfts;
+    }
+
+    public void updateToken(TokenUpdateTransactionBody op, long now) {
+        final var tokenID = Id.fromGrpcToken(op.getToken()).asGrpcToken();
+        validateFalse(tokenID == MISSING_TOKEN, INVALID_TOKEN_ID);
+        if (op.hasExpiry()) {
+            validateFalseOrRevert(validator.isValidExpiry(op.getExpiry()), INVALID_EXPIRATION_TIME);
```

Comment:
I think the expiry should be valid, so `validateTrueOrRevert` would be a better candidate for this check.